### PR TITLE
Network V2: track DAG state with trees

### PR DESCRIPTION
**Pull request by gerardsn**
this PR adds
- tracking of DAG state in xor & iblt trees.
- persistent and transactional storage for the trees
- missing copyright notices

I could use some feedback on what to test in `state` and the TODO in `bbolt_tree`

### CODE REVIEW

File: network/dag/bbolt_tree.go [hunk 1]
```diff
@@ -1,38 +1,143 @@
...
+		})
+		if err != nil {
+			log.Logger().Errorf("failed to add transaction to %s: %s", store.bucketName, err.Error())
+		}
+	}
+}
+
+// migrate builds a tree by walking over the dag and adding all Transaction references to the tree without checking for validity.
+// The tree is stored on disk once it is in sync with the dag.
+// Should only be called on an empty tree.
```
Comment (reinkrul) [inline network/dag/bbolt_tree.go:106]: There's a check for that (and it's always called by Start)

File: network/dag/bbolt_tree.go [hunk 2]
```diff
@@ -1,38 +1,143 @@
...
+// Returns an error the bucket does not exist, or if data in the bucket doesn't match the tree's Data prototype.
+func (store *bboltTree) read(ctx context.Context) error {
+	return storage.BBoltTXUpdate(ctx, store.db, func(_ context.Context, tx *bbolt.Tx) error {
 		// get bucket
-		bucket := tx.Bucket([]byte(bucketName))
+		bucket := tx.Bucket([]byte(store.bucketName))
 		if bucket == nil {
-			return fmt.Errorf("bucket '%s' not found", bucketName)
+			// should only happen once for a new tree/bucket.
+			log.Logger().Warnf("tree bucket '%s' does not exist", store.bucketName)
```
Comment (reinkrul) [inline network/dag/bbolt_tree.go:139]: how is this log useful?

File: network/dag/tree/tree.go [hunk 3]
```diff
@@ -14,21 +33,29 @@ type Data interface {
...
 type Tree interface {
 	// Insert a transaction reference at the specified clock value.
 	// The result of inserting the same ref multiple times is undefined.
-	Insert(ref hash.SHA256Hash, clock uint32) error
+	Insert(ref hash.SHA256Hash, clock uint32)
+	// InsertGetDirty inserts a transaction reference like Insert and also returns the dirty leaves like GetUpdates.
+	// This is an atomic operation to make sure no ResetUpdate can happen in between from calling logic.
+	InsertGetDirty(ref hash.SHA256Hash, clock uint32) map[uint32][]byte
+	// Delete a transaction reference without checking if ref is in the Tree
+	Delete(ref hash.SHA256Hash, clock uint32)
```
Comment (reinkrul) [inline network/dag/tree/tree.go:58]: >without checking if ref is in the Tree

What happens if it's not present? If nothing happens, you can leave ouyt this comment imo

File: network/dag/tree/tree.go [hunk 4]
```diff
@@ -101,29 +132,49 @@ func (t *tree) Load(leaves map[uint32][]byte) error {
 		if err != nil {
 			return err
 		}
-		err = t.updateOrCreatePath(split, func(n *node) error {
-			return n.data.Add(data)
+		t.updateOrCreatePath(split, func(n *node) {
+			n.data.Add(data)
```
Comment (reinkrul) [inline network/dag/tree/tree.go:136]: error is ignored?

**Review by reinkrul — CHANGES_REQUESTED** [review]
Is there some way the tests (preferably integration-level tests) can assert the correctness of the implementation? Say, do some edge case TXs and check whether all nodes have the same state (starting at the `xor` value of the tree)?

File: network/dag/bbolt_tree.go (hunk 2 above)
Comment (gerardsn) [reply network/dag/bbolt_tree.go:139]: the bucket should always exist, except during migration. Maybe change this back to return an error in V3?

File: network/dag/tree/tree.go (hunk 3 above)
Comment (gerardsn) [reply network/dag/tree/tree.go:58]: `Delete` puts -ref in the tree if ref was not in the tree. Deleting a ref from an Xor tree that is not in the tree has the same as effect as `Insert`ing it.

File: network/dag/tree/tree.go (hunk 4 above)
Comment (gerardsn) [reply network/dag/tree/tree.go:136]: No error can occur here since `n.data` and `data` are of the same concrete type.

changed to explicitly ignore error.

File: network/dag/bbolt_tree.go [hunk 5]
```diff
@@ -1,38 +1,143 @@
...
+				<-c.Done()
+				err := c.Err()
+				if err == context.DeadlineExceeded {
+					log.Logger().Warnf("deadline exceeded - rollback transaction %s from %s", transaction.Ref(), store.bucketName)
+					store.tree.Delete(transaction.Ref(), transaction.Clock())
+				}
+			}()
+			tx.OnCommit(func() {
+				store.tree.ResetUpdate()
+				cancel()
```
[the remark this answers is not in the document]
Comment (gerardsn) [reply network/dag/bbolt_tree.go:93]: The cancel() will Delete() the transaction from the tree and (eventually from) its storage, so the tree will not be corrupt. 

However, the tree will now be out of sync with the dag. Since the tree is used for set reconciliation, the node should request the 'missing' transaction from its peers, but the Tx will be dropped (not offered to the dag observers again) since it is already in the dag. 

It is even possible that both trees and the dag in `state` all have a different state.

Any suggestions on what to do with this?
- We could add metrics to log this?
- How likely is this to happen with the current timeout of 10 sec?

any thoughts on this @woutslakhorst

**Comment by gerardsn** [comment]
> Is there some way the tests (preferably integration-level tests) can assert the correctness of the implementation? Say, do some edge case TXs and check whether all nodes have the same state (starting at the `xor` value of the tree)?

I'm happy to write some tests if you can help me get started.

**Comment by gerardsn** [comment]
> > > Is there some way the tests (preferably integration-level tests) can assert the correctness of the implementation? Say, do some edge case TXs and check whether all nodes have the same state (starting at the `xor` value of the tree)?
> > 
> > 
> > I'm happy to write some tests if you can help me get started.
> 
> There's an integration test with paging where you can add a check on the diagnostics?

Added the test as suggested.

**Review by reinkrul — APPROVED** [review]
